Sources/StreamChat/Workers/MessageUpdater.swift

```diff
@@ -255,15 +255,57 @@ class MessageUpdater: Worker {
         messageId: MessageId,
         completion: ((Error?) -> Void)? = nil
     ) {
+        var reaction: MessageReactionDTO?
         let endpoint: Endpoint<EmptyResponse> = .addReaction(
             type,
             score: score,
             enforceUnique: enforceUnique,
             extraData: extraData,
             messageId: messageId
         )
-        apiClient.request(endpoint: endpoint) {
-            completion?($0.error)
+
+        database.write { session in
+            do {
+                reaction = try session.addReaction(
+                    to: messageId,
+                    type: type,
+                    score: score,
+                    extraData: extraData
+                )
+                completion?(nil)
+            } catch {
+                log.error("Failed to add the reaction to the database: \(error)")
+                completion?(error)
+            }
+
+            // if session.addReaction failed we fire the API call, worst case the call with have no effect
+            guard let reaction = reaction else {
+                self.apiClient.request(endpoint: endpoint) { _ in }
+                return
+            }
+
+            reaction.localState = .sending
+
+            // TODO: verify that this is OK and that does not cause problems (api call inside db write)
+            self.apiClient.request(endpoint: endpoint) { result in
+                guard let error = result.error else {
+                    return
+                }
+
+                guard ClientError.isEphemeral(error: error) else {
+                    self.database.write { session in
+                        guard let reaction = try? session.removeReaction(from: messageId, type: type) else {
+                            return
+                        }
+                        reaction.localState = .sendingFailed
+                    }
+                    return
+                }
+
+                self.database.write { _ in
+                    reaction.localState = .pendingSend
+                }
```

Comment:
Will we have a retry button to retry the reaction? Does that UI exist? 🤔 I thought that, if the reaction failed, for whatever reason, it would simply remove it. For example, Twitter and Instagram etc do this, if there is no network, or the request fails, the reaction is simply removed.

Reply:
IMO its a better user experience if we retry behind the scenes, this way you can react to something while offline

Reply:
Hum, how will that work tho? The only way to make it work offline is to save the request in the DB and use a queue. Currently, we have something +/- like this with the `MessageSender` which is our background workers. It is very unlikely that if you are offline for some reason, that in the time it takes to do the retries, you will be able to make the request. So at most this retry mechanism will be useful for connection stability issues, not offline

Reply:
retries are performed forever if the request fails due to connectivity problems.